[terra-pills]Terra filter pills - New Component (#1490)
cerner/terra-framework: packages/terra-framework-docs/src/terra-dev-site/doc/pills/AccessibilityGuide.2.doc.mdx

import { Badge } from 'terra-pills/package.json?dev-site-package';
import { Notice } from '@cerner/terra-docs';

<Badge />

# Accessibility Guide for Terra Filter Pills

> Filters Pills have special interactons and construction that makes them unique from them standard buttons and interactive list items.

Reviewer: Do you mean
```suggestion
Filters Pills have special interactons and construction that makes them unique from standard buttons and interactive list items.
```

> Filters Pills have special interactons and construction that makes them unique from them standard buttons and interactive list items.

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

## Important Considerations:

### Group Label / Heading

>  - Filter Pills as a group requires that an aria-label be provided as accesibility minimum standard protection, but is not visible, and therefore is not
> reccomended to be used alone.
>  - For accessbility best practices, it is reccommended consumers should always include a visible label by using the `ariaLabelledBy` prop.

Reviewer: ```suggestion
 - Filter Pills as a group requires that an `aria-label` be provided as accessibility minimum standard protection, but not visible, and therefore not
reccomended to be used alone.
 - For accessibility best practices, it is recommended that consumers should always include a visible label by using the `ariaLabelledBy` prop.
```

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/c91d230ede1c25913161803525d9462219f5419f)

### Construction

Filter Pills are commonly used in a group or list, which must include a label to indicate the intent or purpose of the contents of the list.
Accordingly, the filter pills component requires the `ariaLabel` prop for accessibility minimum standards, to give a screenreader user the proper
context of the list, but it is not visible to users.

<Notice variant="important" ariaLevel="1">
Accessibility Considerations:

> For best practice, it is recommended that pills include a visible label, which may need to be styled and positioned according to a specific

Reviewer 2: removed new lines and kept it similar to aggregator [here](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

usage and should be done so by each consumer. Use the `ariaLabelledBy` prop in Filter Pills, and provide it the ʻidʼ of the visible label or
heading, which will allow for the visible text to be accessibly connected to the pills.

```diff
    <div className={cx('filter-group-panel')}>
+       <span id="search-terms-fiter-label" className={cx('title-label')}>
+       Patient Education Search Terms:
+       </span>
        <FilterPills
+           ariaLabelledBy="search-terms-fiter-label"
            ariaLabel="Patient Education Search Terms"
            onRemove={handleOnRemove}
        >
            {pills.map((pill, index) => (
                <FilterPills.Pill
                    label={pill.label}
                    labelCategory={pill.labelCategory}
                    id={pill.id}
                    key={pill.id}
                    pillKey={pill.label}
                    metaData={{ index }}
                />
            ))}
        </FilterPills>
    </div>
```

> Note: _see the prop reccomended to set the visible text and the `ariaLabel` prop to the same string values for fallback

Reviewer 2: updated with [commit](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

> protection, even though the `ariaLabelledBy` should override the `ariaLabel` use._

Reviewer 2: updated with [commit](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

</Notice>

## Usability Expectations:

### Interaction Details:

- Filter Pills recieves focus as a group by `tab` keypress, then arrow keys change focus between individual pills within the group.
- Only one focused pill per set of Filter Pills.
- The remove button must not be a focusable element, but can be clickable.
- Removing a filter pill with a keyboard is performed with the `Backspace` or `Delete` key when a pill has focus.

## Keyboard Navigation:
- Filter Pills will have the following keyboard navigation expectations:

<table style={{ margin: "5px", border: "1px solid #0000", 'border-collapse': "collapse" }}>
  <tr>
    <th>Key / Sequence</th>
    <th>Description</th> 
  </tr>
  <tr>

>     <td>`Tab`</td>

Reviewer 2: removed backticks [here](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

>     <td>`Tab`</td>

Reviewer 2: removed

  <tr>
    <td>`Right`</td>
    <td>Right arrow key will move focus to the next pill in the list. When on the last pill in the list, focus remains on the last
    pill and does not cycle to the first pill.</td>
  </tr>
  <tr>
    <td>`Left`</td>
    <td>Left arrow key will move focus to the previous pill in the list. When on the first pill in the list, focus remains on the
    first pill and does not cycle to the last pill.</td>
  </tr>
  <tr>
    <td>`Backspace` or `Delete`</td>
    <td>Removes the pill that currently has focus. After removing, focus should then be placed on the previous pill in the
    list. If no pills remain, focus is temporarily on the Filter Pills group until user navigates away, but cannot be focused
    on again while no interactive pills remain in the Filter Pills group.</td>
  </tr>
  <tr>
    <td>`Enter` or `Space`</td>
    <td>If a pill is selectable via having truncated label, Enter or Space will show the popup to display the full pill label.</td>
  </tr>
</table>

## Support Compliance

### Supported Features & Technology
- Keyboard Interactions
- PC: JAWS Support with Chrome (latest)
- MAC: VoiceOver Support with Chrome (latest), Safari (latest)

> ### Requiring Futher Enhancement to Support

Reviewer 2: updated with [commit](https://github.com/cerner/terra-framework/pull/1490/commits/5a9ff7737735bc1225eab4348ace400f8b5db929)

import { Badge } from 'terra-pills/package.json?dev-site-package';
import { Notice } from '@cerner/terra-docs';

<Badge />

# Accessibility Guide for Terra Filter Pills

Filter Pills have special interactions and construction that makes them unique from standard buttons and interactive list items.

## Important Considerations:

### Construction of Group Label / Heading

- Filter Pills as a group requires that an `aria-label` be provided as accessibility minimum standard protection, but is not visible, and therefore is not reccomended to be used alone.
- For accessibility best practices, it is recommended that consumers should always include a visible label by using the `ariaLabelledBy` prop.
- Filter Pills are commonly used in a group or list, which must include a label to indicate the intent or purpose of the contents of the list. Accordingly, the filter pills component requires the `ariaLabel` prop for accessibility minimum standards, to give a screenreader user the proper context of the list, but it is not visible to users.

<Notice variant="important" ariaLevel="1">
Accessibility Considerations:

> For best practice, it is recommended that pills include a visible label, which may need to be styled and positioned according to a specific usage and should be done so by each consumer. Use the `ariaLabelledBy` prop in Filter Pills, and provide it the ʻidʼ of the visible label or heading, which will allow for the visible text to be accessibly connected to the pills.

Reviewer: ```suggestion
For best practice, it is recommended that pills include a visible label, which may need to be styled and positioned  by the consumer according to their specific usage. Use the `ariaLabelledBy` prop in Filter Pills and provide it the ʻidʼ of the visible label or heading. This will allow for the visible text to be accessibly connected to the pills.
```

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/8d775bb7b611089a8d939625bd2832781544462a#diff-c0f356f3f58def9dbe21de651bee38f73c27a6b729e3d1170e20542793c04538)

```diff
    <div className={cx('filter-group-panel')}>
+       <span id="search-terms-fiter-label" className={cx('title-label')}>
+       Patient Education Search Terms:
+       </span>
        <FilterPills
+           ariaLabelledBy="search-terms-fiter-label"
            ariaLabel="Patient Education Search Terms"
            onRemove={handleOnRemove}
        >
            {pills.map((pill, index) => (
                <Pill
                    label={pill.label}
                    labelCategory={pill.labelCategory}
                    id={pill.id}
                    key={pill.id}
                    metaData={{ index }}
                />
            ))}
        </FilterPills>
    </div>
```

Note: _The visible label referenced by the `ariaLabelledBy` prop and the `ariaLabel` prop should be set to the same string values for fallback protection, even though `ariaLabelledBy` should override `ariaLabel`._
</Notice>

## Usability Expectations:

### Interaction Details:

> - Filter Pills recieves focus as a group by `tab` keypress, then arrow keys change focus between individual pills within the group.

Reviewer: ```suggestion
- Filter Pills receives focus as a group by `tab` keypress, with arrow keys changing focus between individual pills within the group.
```

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/8d775bb7b611089a8d939625bd2832781544462a#diff-c0f356f3f58def9dbe21de651bee38f73c27a6b729e3d1170e20542793c04538)

> - Only one focused pill per set of Filter Pills.

Reviewer: ```suggestion
- Only one pill can be focuesd per set of Filter Pills.
```

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/8d775bb7b611089a8d939625bd2832781544462a#diff-c0f356f3f58def9dbe21de651bee38f73c27a6b729e3d1170e20542793c04538)

- The remove button must not be a focusable element, but can be clickable.
- Removing a filter pill with a keyboard is performed with the `Backspace` or `Delete` key when a pill has focus.

## Keyboard Navigation:
- Filter Pills will have the following keyboard navigation expectations:

<table style={{ margin: "5px", border: "1px solid #0000", 'border-collapse': "collapse" }}>
  <tr>
    <th>Key / Sequence</th>
    <th>Description</th> 
  </tr>
  <tr>
    <td>Tab</td>
    <td>Initial focus goes to the first pill in the list, pressing tab again moves focus out of the Filter Pills group and to the next focusable element on the page.</td>
  </tr>
  <tr>
    <td>Right</td>
    <td>Right arrow key will move focus to the next pill in the list. When on the last pill in the list, focus remains on the last pill and does not cycle to the first pill.</td>
  </tr>
  <tr>
    <td>Left</td>
    <td>Left arrow key will move focus to the previous pill in the list. When on the first pill in the list, focus remains on the first pill and does not cycle to the last pill.</td>
  </tr>
  <tr>
    <td>Backspace or Delete</td>

>     <td>Removes the pill that currently has focus. After removing, focus should then be placed on the previous pill in the list. If no pills remain, focus is temporarily on the Filter Pills group until user navigates away, but cannot be focused on again while no interactive pills remain in the Filter Pills group.</td>

Reviewer: ```suggestion
    <td>Removes the pill that currently has focus. After removal, the focus should then be placed on the previous pill in the list. If no pills remain then the Filter Pills group will be focused until user navigates away. However the Filter Pills group cannot be focused on again while no interactive pills remain.</td>
```

Reviewer 2: [updated](https://github.com/cerner/terra-framework/pull/1490/commits/8d775bb7b611089a8d939625bd2832781544462a#diff-c0f356f3f58def9dbe21de651bee38f73c27a6b729e3d1170e20542793c04538)
